File: software/raspberry_pi/src/navegation/navegation_4.0/AUVError.py
from time import asctime

class AUVError(Exception):
    """
    Generalization of errors or errors that are not identified
    """

    def __init__(self, time = asctime(), type = "Unidentified error", error_code = 95):
        self.time_error = time
        self.type_error = type
        self.error_code = error_code
        super().__init__(f"[{time}] (Error {error_code}): {type}")
        self.log_error()
    
    def log_error(self):
        """
        saves errors to a log file for later checking
        """

        with open("auv_errors.log", "a") as log_file:
            log_file.write(f"[{self.time_error}] (Error {self.error_code}): {self.type_error}\n")

class CollisionDetected(AUVError):
    """
    Detects AUV collisions
    """

    def __init__(self, acceleration = None, time = asctime()):
        self.acceleration = acceleration
        super().__init__(time, type = f"Collision detected based in sensor data: {acceleration} m/s²", error_code = 301)

class HighTempError(AUVError):
    """
    Generalization of temperature error or unidentified temperature problem
    """

    def __init__(self, time = asctime(), temperature = None, type = "High temperature detected", error_code = 95):
        self.temperature = temperature
        super().__init__(time, type, error_code)

class PixhawkHighTempError(HighTempError):
    """
    High temperature at pixhawk
    """

    def __init__(self, temperature = None, time = asctime()):
        super().__init__(time, temperature, type = f"Pixhawk temperature above threshold: {temperature} ºC", error_code = 332)

class RaspberryHighTempError(HighTempError):
    """
    High temperature in raspberry
    """
    
    def __init__(self, temperature = None, time = asctime()):
        super().__init__(time, temperature, type = f"Raspberry temperature above threshold: {temperature} ºC", error_code = 141)

File: software/raspberry_pi/src/navigation/navigation_5.0_mpu6050/src/auverror.cpp
#include "auverror.h"

AUVError::AUVError(string type, int code) : time(getTime()), type(type), code(code){
    logError();
}

void AUVError::logError(){
    ofstream logFile("auv_errors.log", ios::app);
    if(logFile.is_open()){
        logFile << "[" << this->time << "]" << " (Error " << this->code << "): " << this->type << endl;
        logFile.close();
    } else {
        cout << "Failed to open log file: auv_errors.log" << endl;
    }
}

FailedInitializationSensors::FailedInitializationSensors() : AUVError("Failed to create the sensors object.", 151){}

// YOLOCTRL



FailedInitializationYolo ::FailedInitializationYolo(string type, int code) : AUVError(type, code){}

ErrorCreatingSocket::ErrorCreatingSocket() : FailedInitializationYolo("Failed to create the socket.", 252){}

ErrorBindingSocket::ErrorBindingSocket() : FailedInitializationYolo("Failed to bind the socket.", 253){}

ErrorListening::ErrorListening() : FailedInitializationYolo("Failed to listen to the socket.", 254){}

ErrorAcceptingConnection::ErrorAcceptingConnection() : FailedInitializationYolo("Failed to accept the connection.", 255){}

// END YOLOCTRL

CollisionDetected::CollisionDetected(array<double, 3> acceleration) : AUVError("Collision detected based in sensor data: {" + to_string(acceleration[0]) +
                ", " + to_string(acceleration[1]) + ", " + to_string(acceleration[2]) + "} m/s²", 301), acceleration(acceleration){}

array<double, 3> CollisionDetected::getAcceleration(){
    return this->acceleration;
}

FailedConnectThrusters::FailedConnectThrusters() : AUVError("Failed to connect thrusters", 441){}

HighTemperatureError::HighTemperatureError(double temperature, string type, int code) : AUVError(type, code), temperature(temperature){}

PixhawkHighTemperature::PixhawkHighTemperature(double temperature) : HighTemperatureError(temperature, "Pixhawk temperature above threshold " + to_string(temperature) + "ºC", 332){}

RaspberryHighTemperature::RaspberryHighTemperature(double temperature) : HighTemperatureError(temperature, "Raspberry temperature above threshold " + to_string(temperature) + "ºC", 141){}

FailedFiringTorpedo::FailedFiringTorpedo() : AUVError("Failed to fire torpedo", 820){}